### PR TITLE
NASA Marine Debris dataset

### PR DESCRIPTION
**Pull request by isaaccorley**
This PR adds the [NASA Marine Debris](https://mlhub.earth/data/nasa_marine_debris) dataset, a dataset for object detection of marine debris in 3m RGB 256x256 Planet Labs PlanetScope imagery. Labels come in geojson and numpy (.npy) bbox xyxy formats (I'm loading from numpy for speed).

Notes:
Each image can have a variable number of "boxes" for objects. This becomes problematic when loading a batch of boxes in a datamodule because they can't be stacked. It's common practice in [torchvision](https://pytorch.org/vision/stable/_modules/torchvision/models/detection/faster_rcnn.html) and [detectron2](https://github.com/facebookresearch/detectron2) to get around this by loading a list of sample dicts instead of a single batch dict and then the model handles the list internally. The easy workaround is to just create a custom `collate_fn` in the datamodule. I'm open to other suggestions though.

Examples:

image
image

### CODE REVIEW

File: torchgeo/datasets/nasa_marine_debris.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+from .geo import VisionDataset
+from .utils import dataset_split, download_radiant_mlhub_dataset, extract_archive
+
+# https://github.com/pytorch/pytorch/issues/60979
+# https://github.com/pytorch/pytorch/pull/61045
+DataLoader.__module__ = "torch.utils.data"
+
+
+def collate_fn(batch: List[Dict[str, Tensor]]) -> Dict[str, Any]:
```
Comment (adamjstewart) [inline torchgeo/datasets/nasa_marine_debris.py:27]: I think this is equivalent to and can be replaced by `torchgeo.datasets.utils.stack_samples`

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: torchgeo/datasets/nasa_marine_debris.py (hunk 1 above)
Comment (isaaccorley) [reply torchgeo/datasets/nasa_marine_debris.py:27]: It's not. `stack_samples` just stacks each samples images, masks, etc. in a batch together. Each image in this dataset can have a variable number of boxes so you can't stack them together (this is common for other object detection datasets as well). See my notes in the description. We could probably make a custom `stack_box_samples` or something in the future to reuse.

**Review by adamjstewart — APPROVED** [review]

File: tests/datasets/test_nasa_marine_debris.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    ) -> None:
+        shutil.rmtree(dataset.root)
+        os.makedirs(str(tmp_path), exist_ok=True)
+        Dataset().download(output_dir=str(tmp_path))
+        print(os.listdir(str(tmp_path)))
+        NASAMarineDebris(root=str(tmp_path), download=False)
+
+    def test_not_downloaded(self, tmp_path: Path) -> None:
+        err = rf"Dataset not found in `root={str(tmp_path)}` directory and "
+        "`download=False`, either specify a different `root` directory"
```
Comment (adamjstewart) [inline tests/datasets/test_nasa_marine_debris.py:74]:
```suggestion
        "`download=False`, either specify a different `root` directory "
```

**Review by adamjstewart — APPROVED** [review]

**Review by adamjstewart — APPROVED** [review]
